optaplanner-website-root/content/blog/2022/03/04/Real-time planning meets the SolverManager.adoc

```diff
@@ -0,0 +1,197 @@
+
+[source,java]
+----
+public class SolvingService {
+    @Inject
+    SolverManager<CloudBalance, Long> solverManager; // <1>
+
+    public void startSolving(Long problemId) {
+        solverManager.solveAndListen(problemId, this::loadProblem, bestSolution -> saveSolution(problemId, bestSolution)); // <2>
+    }
+
+    public void addComputer(Long problemId, CloudComputer computer) {
+        solverManager.addProblemChange(problemId, (workingSolution, problemChangeDirector) -> { // <3>
+            List<CloudComputer> computerList = new ArrayList<>(workingSolution.getComputerList());
+            workingSolution.setComputerList(computerList);
+            problemChangeDirector.addProblemFact(computer, computerList::add);
+        });
+    }
+
+    private CloudBalance loadProblem(Long problemId) {
+        // Load the input problem identified by the problemId.
+    }
+
+    private void saveSolution(Long problemId, CloudBalance cloudBalance) {
+        // Save the best solution, or send it to UI, etc.
+    }
+}
+----
+<1> Injects the `SolverManager`, assuming the application runs on top of Quarkus. Similarly, your can inject the `SolverManager`
+in a Spring Boot application using the `@Autowired` annotation.
+<2> Submits a problem to the `SolverManager`. Every best solution is passed to the `saveSolution()` method.
+<3> Adds a new computer to the working solution identified by the `problemId` via a `ProblemChange`.
+
+== Testing ProblemChanges
+
+As any piece of a software project that implements non-trivial logic, ``ProblemChange``s should be unit-tested.
+What is the testable contract? First, make sure the right methods on the `ProblemChangeDescriptor` are called, and second,
+the working solution must contain the expected changes.
+
+To help you with testing if the correct methods of the `ProblemChangeDescriptor` were called there is the `MockProblemChangeDirector`
+available in https://search.maven.org/artifact/org.optaplanner/optaplanner-test[org.optaplanner:optaplanner-test].
+The last example below shows how to use the `MockProblemChangeDirector` together with https://site.mockito.org/[Mockito].
+
+[source,java]
+----
+public class CloudBalanceChangeTest {
+    @Test
+    public void addProcess() {
+        CloudProcess newProcess = new CloudProcess();
+        CloudBalance workingSolution = CloudBalance.emptySolution();
+        MockProblemChangeDirector mockProblemChangeDirector = Mockito.spy(new MockProblemChangeDirector()); // <1>
+        ProblemChange problemChange = new AddProcessProblemChange(newProcess);
+        problemChange.doChange(workingSolution, mockProblemChangeDirector); // <2>
+        verify(mockProblemChangeDirector).addEntity(same(newProcess), any()); // <3>
+        assertEquals(1, workingSolution.getProcessList().size()); // <4>
+        assertSame(newProcess, workingSolution.getProcessList().get(0)); // <4>
+    }
+}
+----
+<1> Creates a `Mockito.spy()` object of the `MockProblemChangeDirector`.
```

Comment:
Not sure this explains anything. You're just saying what the code mechanically does.
Is it reasonable to assume people know what `Mockito` is? What `spy()` does?